iroh-bytes/src/store/bao_file.rs

```diff
@@ -0,0 +1,1238 @@
+    sizes: SizeInfo,
+}
+
+/// Keep track of the most precise size we know of.
+///
+/// When in memory, we don't have to write the size for every chunk to a separate
+/// slot, but can just keep the best one.
+#[derive(Debug, Default)]
+pub struct SizeInfo {
+    offset: u64,
+    size: u64,
+}
+
+impl SizeInfo {
+    /// Create a new size info for a complete file of size `size`.
+    fn complete(size: u64) -> Self {
+        let mask = (1 << IROH_BLOCK_SIZE.0) - 1;
+        // offset of the last bao chunk in a file of size `size`
+        let last_chunk_offset = size & mask;
+        Self {
+            offset: last_chunk_offset,
+            size,
+        }
+    }
+
+    /// Write a size at the given offset. The size at the highest offset is going to be kept.
+    fn write(&mut self, offset: u64, size: u64) {
+        // >= instead of > because we want to be able to update size 0, the initial value.
+        if offset >= self.offset {
+            self.offset = offset;
+            self.size = size;
+        }
+    }
+
+    /// Persist into a file where each chunk has its own slot.
+    fn persist(&self, mut target: impl WriteAt) -> io::Result<()> {
+        if self.offset & ((IROH_BLOCK_SIZE.bytes() as u64) - 1) != 0 {
+            return Err(io::Error::new(
+                io::ErrorKind::InvalidData,
+                "offset not aligned",
+            ));
+        }
+        let size_offset = (self.offset >> IROH_BLOCK_SIZE.0) << 3;
+        target.write_all_at(size_offset, self.size.to_le_bytes().as_slice())?;
+        Ok(())
+    }
+
+    /// The current size, representing the most correct size we know.
+    pub fn current_size(&self) -> u64 {
+        self.size
+    }
+
+    /// Convert to a vec in slot format.
+    pub fn to_vec(&self) -> io::Result<Vec<u8>> {
+        let mut res = Vec::new();
+        self.persist(&mut res)?;
+        Ok(res)
+    }
+}
+
+impl MutableMemStorage {
+    /// Get the parts data, outboard and sizes
+    pub fn into_parts(self) -> (SparseMemFile, SparseMemFile, SizeInfo) {
+        (self.data, self.outboard, self.sizes)
+    }
+
+    /// Create a new mutable mem storage from the given data
+    pub fn complete(bytes: Bytes) -> (Self, iroh_base::hash::Hash) {
+        let (outboard, hash) = raw_outboard(bytes.as_ref());
+        let res = Self {
+            data: bytes.to_vec().into(),
+            outboard: outboard.into(),
+            sizes: SizeInfo::complete(bytes.len() as u64),
+        };
+        (res, hash)
+    }
+
+    /// Persist the batch to disk, creating a FileBatch.
+    fn persist(&self, paths: DataPaths) -> io::Result<FileStorage> {
+        let mut data = create_read_write(&paths.data)?;
+        let mut outboard = create_read_write(&paths.outboard)?;
+        let mut sizes = create_read_write(&paths.sizes)?;
+        self.data.persist(&mut data)?;
+        self.outboard.persist(&mut outboard)?;
+        self.sizes.persist(&mut sizes)?;
+        data.sync_all()?;
```

Comment:
nit: wouldn't it make more sense to call these pair wise `.persist()` followed by the matching `sync_all` call?

Reply:
You mean because it will be faster? Yes, might be. You could also do it in parallel, but really I was going to leave optimizations like that for later and just go for hopefully correct code.

Reply:
I was more thinking about error cases than performance, but I am undecided which ordering in an error case is the better one